src/escape.rs

```diff
@@ -131,6 +155,116 @@ where
         }
         i += 1;
     }
-    w.write_all(&bytes[mark..])?;
+    w.write_all(&bytes[mark..])
+}
+
+/// Calls the given function with the index of every byte that is in
+/// [", &, <, >] and for no other byte.
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+fn scan_simd<F>(bytes: &[u8], mut offset: usize, mut callback: F) -> io::Result<()>
+    where F: FnMut(usize) -> io::Result<()>
+{
+    if bytes.is_empty() {
+        // we need this special case for when we are passed an empty subslice
+        // that is exactly on a page boundary. we would try to read on a page
+        // that is potentially not ours and crash
+        return Ok(());
+    }
+
+    let lower_vec = unsafe { _mm_set_epi8( 0, 62, 0, 60, 0, 0, 0, 0, 0, 38, 0, 0, 0, 34, 0, 127 ) };
+    let upperbound = bytes.len().saturating_sub(16);
+
+    while offset < upperbound {
+        let mut mask = unsafe {
+            let raw_ptr = transmute(bytes.as_ptr().offset(offset as isize));
+            let v = _mm_loadu_si128(raw_ptr);
+            let expected = _mm_shuffle_epi8(lower_vec, v);
+            let matches = _mm_cmpeq_epi8(expected, v);
+            _mm_movemask_epi8(matches)
+        };
+
+        while mask != 0 {
+            let ix = mask.trailing_zeros();
+            callback(offset + ix as usize)?;
+            mask ^= mask & -mask;
+        }
+
+        offset += 16;
+    }
+
+    // final iteration. there's two scenario's:
+    //  * the remaining bytes cross a 16-byte aligned chunk: in this case
+    //    we know that the next chunk is on a page assigned to our process
+    //    and it's safe to read (by the assertion that page sizes are multiples
+    //    of 16 bytes)
+    //  * the remaining bytes do not cross a 16-byte aligned chunk: reading
+    //    16 bytes more bytes could cross onto a new page which isn't ours
+    //    and crash the program. However, we since all the bytes are contained
+    //    in the current chunk, we can just read that and shift off the
+    //    bytes we aren't interested in
+    let alignment = (bytes.as_ptr() as usize) % 16;
+    let remaining_bytes = bytes.len() - offset;
+
+    let mut mask = if alignment + remaining_bytes > 16 {
+        unsafe {
+            let raw_ptr = transmute(bytes.as_ptr().offset(offset as isize));
+            let v = _mm_loadu_si128(raw_ptr);
+            let expected = _mm_shuffle_epi8(lower_vec, v);
+            let matches = _mm_cmpeq_epi8(expected, v);
+            _mm_movemask_epi8(matches)
+        }
+    } else {
+        unsafe {
+            let raw_ptr = transmute(bytes.as_ptr().offset(offset as isize - alignment as isize));
+            // yay, aligned load!
+            let v = _mm_load_si128(raw_ptr);
```

Comment:
I don't think I'd bother with an aligned load here. Can you observe a perf difference from this? If you're going to worry about aligned loads, then it should probably be in the loop above, but even then I've found it difficult to observe a difference.

Reply:
I've read that on many modern CPUs there's even absolutely no difference. The aligned load is there for safety reasons (to not cross a 16-byte boundary and no page boundary by extension, at least, that's the idea), not performance. I guess the aligned intrinsic is there just for signalling.

Reply:
Ah gotya.